File: apps/trading-journal/backend/app/schemas/analytics.py
"""
Analytics schemas for API requests and responses.
"""

from __future__ import annotations

from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List
from datetime import date, datetime
from decimal import Decimal


class PerformanceMetrics(BaseModel):
    """Performance metrics response."""
    sharpe_ratio: Optional[Decimal] = Field(None, description="Sharpe ratio")
    sortino_ratio: Optional[Decimal] = Field(None, description="Sortino ratio")
    max_drawdown: Optional[Decimal] = Field(None, description="Maximum drawdown percentage")
    avg_drawdown: Optional[Decimal] = Field(None, description="Average drawdown percentage")
    win_rate: Optional[Decimal] = Field(None, description="Win rate percentage")
    profit_factor: Optional[Decimal] = Field(None, description="Profit factor")
    avg_win: Optional[Decimal] = Field(None, description="Average winning trade")
    avg_loss: Optional[Decimal] = Field(None, description="Average losing trade")
    best_trade: Optional[Decimal] = Field(None, description="Best trade P&L")
    worst_trade: Optional[Decimal] = Field(None, description="Worst trade P&L")
    total_trades: int = Field(0, description="Total number of trades")
    as_of: datetime = Field(default_factory=datetime.now, description="Timestamp of calculation")
    
    model_config = ConfigDict(from_attributes=True)


class TickerPerformance(BaseModel):
    """Performance breakdown by ticker."""
    ticker: str = Field(..., description="Ticker symbol")
    total_trades: int = Field(0, description="Total number of trades")
    net_pnl: Decimal = Field(Decimal("0"), description="Net P&L")
    gross_pnl: Decimal = Field(Decimal("0"), description="Gross P&L")
    win_rate: Optional[Decimal] = Field(None, description="Win rate percentage")
    profit_factor: Optional[Decimal] = Field(None, description="Profit factor")
    avg_win: Optional[Decimal] = Field(None, description="Average winning trade")
    avg_loss: Optional[Decimal] = Field(None, description="Average losing trade")
    winners: int = Field(0, description="Number of winning trades")
    losers: int = Field(0, description="Number of losing trades")
    
    model_config = ConfigDict(from_attributes=True)


class TickerPerformanceResponse(BaseModel):
    """Response for ticker performance breakdown."""
    date_from: Optional[date] = Field(None, description="Start date filter")
    date_to: Optional[date] = Field(None, description="End date filter")
    tickers: List[TickerPerformance] = Field(default_factory=list, description="Performance by ticker")
    as_of: datetime = Field(default_factory=datetime.now, description="Timestamp of calculation")
    
    model_config = ConfigDict(from_attributes=True)


class TypePerformance(BaseModel):
    """Performance breakdown by trade type."""
    trade_type: str = Field(..., description="Trade type (STOCK, OPTION, etc.)")
    total_trades: int = Field(0, description="Total number of trades")
    net_pnl: Decimal = Field(Decimal("0"), description="Net P&L")
    gross_pnl: Decimal = Field(Decimal("0"), description="Gross P&L")
    win_rate: Optional[Decimal] = Field(None, description="Win rate percentage")
    profit_factor: Optional[Decimal] = Field(None, description="Profit factor")
    avg_win: Optional[Decimal] = Field(None, description="Average winning trade")
    avg_loss: Optional[Decimal] = Field(None, description="Average losing trade")
    winners: int = Field(0, description="Number of winning trades")
    losers: int = Field(0, description="Number of losing trades")
    
    model_config = ConfigDict(from_attributes=True)


class TypePerformanceResponse(BaseModel):
    """Response for trade type performance breakdown."""
    date_from: Optional[date] = Field(None, description="Start date filter")
    date_to: Optional[date] = Field(None, description="End date filter")
    types: List[TypePerformance] = Field(default_factory=list, description="Performance by trade type")
    as_of: datetime = Field(default_factory=datetime.now, description="Timestamp of calculation")
    
    model_config = ConfigDict(from_attributes=True)


class PlaybookPerformance(BaseModel):
    """Performance breakdown by playbook/strategy."""
    playbook: str = Field(..., description="Playbook/strategy name")
    total_trades: int = Field(0, description="Total number of trades")
    net_pnl: Decimal = Field(Decimal("0"), description="Net P&L")
    gross_pnl: Decimal = Field(Decimal("0"), description="Gross P&L")
    win_rate: Optional[Decimal] = Field(None, description="Win rate percentage")
    profit_factor: Optional[Decimal] = Field(None, description="Profit factor")
    avg_win: Optional[Decimal] = Field(None, description="Average winning trade")
    avg_loss: Optional[Decimal] = Field(None, description="Average losing trade")
    winners: int = Field(0, description="Number of winning trades")
    losers: int = Field(0, description="Number of losing trades")
    
    model_config = ConfigDict(from_attributes=True)


class PlaybookPerformanceResponse(BaseModel):
    """Response for playbook performance breakdown."""
    date_from: Optional[date] = Field(None, description="Start date filter")
    date_to: Optional[date] = Field(None, description="End date filter")
    playbooks: List[PlaybookPerformance] = Field(default_factory=list, description="Performance by playbook")
    as_of: datetime = Field(default_factory=datetime.now, description="Timestamp of calculation")
    
    model_config = ConfigDict(from_attributes=True)

File: apps/trading-journal/frontend/src/types/analytics.ts
/**
 * TypeScript types for analytics-related data structures.
 * 
 * These types match the Pydantic schemas from the backend.
 */

export interface PerformanceMetrics {
  sharpe_ratio?: number
  sortino_ratio?: number
  max_drawdown?: number
  avg_drawdown?: number
  win_rate?: number
  profit_factor?: number
  avg_win?: number
  avg_loss?: number
  best_trade?: number
  worst_trade?: number
  total_trades: number
  as_of: string // ISO datetime string
}

export interface TickerPerformance {
  ticker: string
  total_trades: number
  net_pnl: number
  gross_pnl: number
  win_rate?: number
  profit_factor?: number
  avg_win?: number
  avg_loss?: number
  winners: number
  losers: number
}

export interface TickerPerformanceResponse {
  date_from?: string
  date_to?: string
  tickers: TickerPerformance[]
  as_of: string
}

export interface TypePerformance {
  trade_type: string
  total_trades: number
  net_pnl: number
  gross_pnl: number
  win_rate?: number
  profit_factor?: number
  avg_win?: number
  avg_loss?: number
  winners: number
  losers: number
}

export interface TypePerformanceResponse {
  date_from?: string
  date_to?: string
  types: TypePerformance[]
  as_of: string
}

export interface PlaybookPerformance {
  playbook: string
  total_trades: number
  net_pnl: number
  gross_pnl: number
  win_rate?: number
  profit_factor?: number
  avg_win?: number
  avg_loss?: number
  winners: number
  losers: number
}

export interface PlaybookPerformanceResponse {
  date_from?: string
  date_to?: string
  playbooks: PlaybookPerformance[]
  as_of: string
}

